lib/veri.ml

```diff
@@ -0,0 +1,261 @@
+    | Ok r -> match r with
+      | mem', Some insn, `finished -> Ok (mem',insn)
+      | _, None, _ -> 
+        let er = Error.of_string "nothing was disasmed" in
+        Error (`Disasm_error er)
+      | _, _, `left _ -> Error `Overloaded_chunk
+
+  let insn_name = Dis.Insn.name 
+end
+
+module Events = Value.Set
+
+class context report trace = object(self:'s)
+  inherit Veri_traci.context trace as super
+  val report : Veri_report.t = report
+  val events = Events.empty
+  val other = None
+  val descr : string option = None
+  val error : error option = None
+
+  method merge: 's =
+    let report =
+      match error with 
+      | Some er -> Veri_report.notify report er
+      | None ->
+        let events = Option.(value_exn self#other)#events in
+        let events' = self#events in
+        let left = Set.diff events events' in
+        let right = Set.diff events' events in
+        let frame = Veri_report.({left;right}) in
+        match descr with
+        | None -> report
+        | Some name -> Veri_report.update report name frame in   
+    {<other = None; error = None; descr = None; 
+      events = Events.empty; report = report >}
+
+  method replay =
+    let new_i = Option.value_exn other in
+    new_i#backup self
+
+  method events: Events.t = events
+  method split = self#backup self
+  method other = other
+  method report = report
+  method set_description s = {<descr = Some s >}
+  method register_event ev = {< events = Set.add events ev >}
+  method notify_error er = {<error = Some er >}
+  method backup someone = {<other = Some someone; events = Events.empty >}
+end
+
+let target_info arch = 
+  let module Target = (val target_of_arch arch) in
+  Target.CPU.mem, Target.lift
+
+let memory_of_chunk endian chunk = 
+  Bigstring.of_string (Chunk.data chunk) |>
+  Memory.create endian (Chunk.addr chunk) 
+
+let other_events c = match c#other with 
+  | None -> []
+  | Some c -> Set.to_list c#events
+
+class ['a] t arch dis is_interesting =
+  let endian = Arch.endian arch in
+  let mem_var, lift = target_info arch in
+
+  object(self)
+    constraint 'a = #context
+    inherit ['a] Veri_traci.t arch as super
+
+    method private update_event ev =
+      if is_interesting ev then SM.update (fun c -> c#register_event ev)
+      else SM.return () 
+
+    method private resolve_var : var -> 'a r = fun var ->
+      super#lookup var >>= fun r ->
+      match value r with 
+      | Bil.Mem _ | Bil.Imm _ -> SM.return r
+      | Bil.Bot -> 
+        SM.get () >>= fun ctxt ->
+        let same_var mv = Move.cell mv = var in
+        match find_reg_read (other_events ctxt) same_var with
+        | None -> super#lookup var
+        | Some mv -> self#eval_exp (Bil.int (Move.data mv))
+
+    method private emit_reg_read var data : ' u = 
+      self#update_event (create_reg_read var data)
+
+    method! lookup var : 'a r =
+      self#resolve_var var >>= fun r ->
+      if is_imm_var var then
```

Comment:
why do you need this `is_imm_var` check? Why just don't `match imm_result (value r)` and if it is `Some thing` then it imples that it is `Imm`. Moreover, maybe the `imm_result` helper is not needed, it introduces more noise, then solves. Just `match value result with` would be nicer.